front/lib/resources/membership_resource.ts

```diff
@@ -0,0 +1,374 @@
+    });
+    if (memberships.length === 0) {
+      return null;
+    }
+    if (memberships.length > 1) {
+      logger.error(
+        {
+          panic: true,
+          userId,
+          workspaceId: workspace.id,
+          memberships,
+        },
+        "Unreachable: Found multiple latest memberships for user in workspace."
+      );
+      throw new Error(
+        `Unreachable: Found multiple latest memberships for user ${userId} in workspace ${workspace.id}`
+      );
+    }
+    return memberships[0];
+  }
+
+  static async getMembersCountForWorkspace({
+    workspace,
+    activeOnly,
+    transaction,
+  }: {
+    workspace: LightWorkspaceType;
+    activeOnly: boolean;
+    transaction?: Transaction;
+  }): Promise<number> {
+    const where: WhereOptions<InferAttributes<MembershipModel>> = activeOnly
+      ? {
+          endAt: {
+            [Op.or]: [{ [Op.eq]: null }, { [Op.gt]: new Date() }],
+          },
+          startAt: {
+            [Op.lte]: new Date(),
+          },
+        }
+      : {};
+
+    where.workspaceId = workspace.id;
+
+    return MembershipModel.count({
+      where,
+      distinct: true,
+      col: "userId",
+      transaction,
+    });
+  }
+
+  static async createMembership({
+    userId,
+    workspace,
+    role,
+    startAt = new Date(),
+    transaction,
+  }: {
+    userId: number;
+    workspace: LightWorkspaceType;
+    role: MembershipRoleType;
+    startAt?: Date;
+    transaction?: Transaction;
+  }): Promise<MembershipResource> {
+    if (startAt > new Date()) {
+      throw new Error("Cannot create a membership in the future");
+    }
+    if (
+      await MembershipModel.count({
+        where: {
+          userId,
+          workspaceId: workspace.id,
+          endAt: {
+            [Op.or]: [{ [Op.eq]: null }, { [Op.gt]: startAt }],
+          },
+        },
+        transaction,
+      })
+    ) {
+      throw new Error(
+        `User ${userId} already has an active membership in workspace ${workspace.id}`
+      );
+    }
+    const newMembership = await MembershipModel.create(
+      {
+        startAt,
+        userId,
+        workspaceId: workspace.id,
+        role,
+      },
+      { transaction }
+    );
+
+    return new MembershipResource(MembershipModel, newMembership.get());
+  }
+
+  static async revokeMembership({
+    userId,
+    workspace,
+    endAt = new Date(),
+    transaction,
+  }: {
+    userId: number;
+    workspace: LightWorkspaceType;
+    endAt?: Date;
+    transaction?: Transaction;
+  }): Promise<
+    Result<
+      undefined,
+      {
+        type: "not_found" | "already_revoked";
+      }
+    >
+  > {
+    const membership = await this.getLatestMembershipOfUserInWorkspace({
+      userId,
+      workspace,
+      transaction,
+    });
+    if (!membership) {
+      return new Err({ type: "not_found" });
+    }
+    if (endAt < membership.startAt) {
+      throw new Error("endAt must be after startAt");
+    }
+    if (membership.endAt) {
```

Comment:
No currently indeed, but would be relatively straight forward to add support for. Wanted to avoid introducing such capability without a clear need for it.